cmd/agent/app/builder.go

```diff
@@ -74,20 +75,13 @@ type Builder struct {
 	Processors     []ProcessorConfiguration    `yaml:"processors"`
 	SamplingServer SamplingServerConfiguration `yaml:"samplingServer"`
 
-	// CollectorHostPorts are host:ports of a static list of Jaeger Collectors.
-	CollectorHostPorts []string `yaml:"collectorHostPorts"`
-
-	// MinPeers is the min number of servers we want the agent to connect to.
-	// If zero, defaults to min(3, number of peers returned by service discovery)
-	DiscoveryMinPeers int `yaml:"minPeers"`
+	// These 3 fields are copied from reporter.Builder because yaml does not parse embedded structs
+	CollectorHostPorts   []string `yaml:"collectorHostPorts"`
+	DiscoveryMinPeers    int      `yaml:"minPeers"`
+	CollectorServiceName string   `yaml:"collectorServiceName"`
 
-	// CollectorServiceName is the name that Jaeger Collector's TChannel server
-	// responds to.
-	CollectorServiceName string `yaml:"collectorServiceName"`
+	tchreporter.Builder
```

Comment:
I had to deal with two issues. 
  1. this Builder does a double duty as yaml-enabled configuration that our Puppet-deployed agents depend on, so I did not want to break config compatibility by pulling builder out completely
  1. there is a coupling between the main reporter's TChannel and the sampling server that also uses it

I think multiple reporters is still a fairly edge case that can be handled via custom initialization. The cleanest way would be to make the Agent depend on externally provided Reporter and SamplingServer.  But this will break config compatibility.